src/sdi-notify.c

```diff
@@ -287,62 +292,110 @@ void sdi_notify_pending_refresh_one(SdiNotify *self, SnapdSnap *snap) {
+    notify_pending_refresh_forced(self, snap, TRUE);
+    return TRUE;
+  } else if (next_refresh <= TIME_TO_SHOW_ALERT_BEFORE_FORCED_REFRESH) {
+    // If the remaining time is less than this, force a notification.
+    notify_pending_refresh_forced(self, snap, FALSE);
+    return TRUE;
+  }
+  return FALSE;
+}
+
+static gchar *get_name_from_snap(SnapdSnap *snap) {
+  const gchar *name = snapd_snap_get_name(snap);
+  g_autoptr(GAppInfo) app_info = sdi_get_desktop_file_from_snap(snap);
+  if (app_info != NULL) {
+    const gchar *name2 = g_app_info_get_display_name(app_info);
+    if (name2 != NULL) {
+      name = name2;
+    }
+  }
+  return g_strdup(name);
+}
+
+void sdi_notify_pending_refresh(SdiNotify *self, GSList *snaps) {
   g_return_if_fail(SDI_IS_NOTIFY(self));
   g_return_if_fail(snaps != NULL);
 
-  g_autoptr(GString) body = g_string_new(_("Close the apps to start updating"));
-  for (; snaps != NULL; snaps = snaps->next) {
-    SnapdSnap *snap = (SnapdSnap *)snaps->data;
-    const gchar *name = snapd_snap_get_name(snap);
-    g_autoptr(GAppInfo) app_info = sdi_get_desktop_file_from_snap(snap);
+  g_autofree gchar *title = NULL;
+  g_autofree gchar *body = NULL;
+  g_autoptr(GAppInfo) app_info = NULL;
+  g_autoptr(GDesktopAppInfo) app_info2 = NULL;
+
+  guint n_snaps = g_slist_length(snaps);
+
+  GIcon *icon = NULL;
+  if (n_snaps == 1) {
+    g_autofree gchar *snap_name = get_name_from_snap((SnapdSnap *)snaps->data);
+    title = g_strdup_printf(_("Update available for %s"), snap_name);
+    body = g_strdup(_("Quit the app to update it now."));
+    app_info = sdi_get_desktop_file_from_snap((SnapdSnap *)snaps->data);
     if (app_info != NULL) {
-      name = g_app_info_get_display_name(app_info);
+      icon = g_app_info_get_icon(app_info);
     }
-
-    GTimeSpan difference = get_remaining_time(snap) / 1000000;
-    if (difference > 86400) {
-      g_string_append_printf(body, _(" (%s %ld days left)"), name,
-                             difference / 86400);
-    } else if (difference > 3600) {
-      g_string_append_printf(body, _(" (%s %ld hours left)"), name,
-                             difference / 3600);
-    } else {
-      g_string_append_printf(body, _(" (%s %ld minutes left)"), name,
-                             difference / 60);
+  } else {
+    // although the case for 1 app is managed outside this, I put it here to
+    // ensure that ngettext works as expected, and to ensure that translators
+    // know what is going on.
+    title = g_strdup_printf(ngettext("Update available for %d app",
+                                     "Updates available for %d apps", n_snaps),
+                            n_snaps);
+    switch (n_snaps) {
+    case 2:
+      body =
+          g_strdup_printf(_("Quit %s and %s to update them now."),
+                          get_name_from_snap((SnapdSnap *)snaps->data),
+                          get_name_from_snap((SnapdSnap *)snaps->next->data));
+      break;
+    case 3:
```

Comment:
What occurs when there are more than 3 snaps?

Reply:
I forgot the `default`. Fixed.